src/triage/component/catwalk/utils.py

```diff
@@ -204,13 +206,20 @@ def save_db_objects(db_engine, db_objects):
 
     Args:
         db_engine (sqlalchemy.engine)
-        db_objects (list) SQLAlchemy model objects, corresponding to a valid table
+        db_objects (iterator) SQLAlchemy model objects, corresponding to a valid table
     """
-    with tempfile.TemporaryFile(mode="w+") as f:
-        writer = csv.writer(f, quoting=csv.QUOTE_MINIMAL)
-        for db_object in db_objects:
+
+    first_object = next(db_objects)
+    type_of_object = type(first_object)
+    def write_output(file_like):
```

Comment:
Just to note, _if you like_, you could define this at the module level, (as a helper like `_write_csv(db_objects, file_like)`), and then pass a `functools.partial` to `PipeTextIO`.

(In that case it might also be best for the helper to take over ensuring that it has an iterator, grabbing the first type, _etc._, just for encapsulation.)

(I was also thinking that the `PipeTextIO` constructor _could_ take **arguments you want to pass to your function, as a nicety; but, I'm not sure we want to complicate that right now.)

Reply:
Note that the type of the first object is also used directly in the postgres_copy, so I don't actually want to encapsulate that. Making the helper module level, I get this:

```
def _write_csv(file_like, db_objects, type_of_object):
    writer = csv.writer(file_like, quoting=csv.QUOTE_MINIMAL)
    for db_object in db_objects:
        if type(db_object) != type_of_object:
            raise TypeError("Cannot copy collection of objects to db as they are not all "
                            f"of the same type. First object was {type_of_object} "
                            f"and later encountered a {type(db_object)}")
        writer.writerow(
            [getattr(db_object, col.name) for col in db_object.__table__.columns]
        )

@db_retry
def save_db_objects(db_engine, db_objects):
    """Saves a collection of SQLAlchemy model objects to the database using a COPY command

    Args:
        db_engine (sqlalchemy.engine)
        db_objects (iterable) SQLAlchemy model objects, corresponding to a valid table
    """
    db_objects = iter(db_objects) 
    first_object = next(db_objects)
    type_of_object = type(first_object)

    with PipeTextIO(partial(
            _write_csv,
            db_objects=chain((first_object,), db_objects),
            type_of_object=type_of_object
    )) as pipe:
        postgres_copy.copy_from(pipe, type_of_object, db_engine, format="csv")
```

Do you think this looks better? I think a functools.partial is its own level of indirection, and although I like to use it I generally only do so when it solves a bigger problem, such as passing around the partially bound function to multiple different places (e.g. catwalk/evaluation.py). I'm less convinced of the utility of simply trading a dynamically scoped named function for a functools.partial, though I'm open to arguments for it.

Reply:
Bah, OK, `type_of_object` :+1: 

Yeah, I think that's much clearer :+1: :+1: 

I agree that there are pros and cons to each, but I think the overriding "pros" to splitting out `_write_csv` are what I mentioned before:

* `_write_csv` doesn't inherit the dynamic scope of `save_db_objects`, (and there's less potential for leakage)
* (and, you're only constructing an object dynamically, not defining a function every time, which is ostensibly better for performance)

Certainly this could be a `lambda` or something rather than a `partial`, but I think a `partial` beats a `lambda` here. This is as much what `partials` are for – as a functional interface adapter – as when partials "must" be used because the signature is being passed around.

I am fond of dynamically-constructed functions for some things, but more when they're on the scale of a lambda, _i.e._ with _very_ little internal complexity.